### PR TITLE
proc: Load balancing, removing dead code, tracking allocated cpu

### PR DESCRIPTION
**Pull request by notYuriy**
This PR
- [x] Removes condvar and semaphore, they won't be used anyway (everything will use messaging instead).
- [x] Moves yield logic in arch-dependent code, so that task can be enqueued on the other core without any problems.  Most shared code is already wrapped anyway, now yield can wait using `mwait`/`wfi` or something
- [x] Adds basic load balancer, implemented with a linear scan
- [x] Wakes up tasks where needed

### CODE REVIEW

File: src/thread/scheduler.zig [hunk 1]
```diff
@@ -1,31 +1,67 @@
...
+const task_alloc = os.memory.vmm.backed(.Ephemeral);
+var balancer_lock = os.thread.Spinlock{};
+
+/// Allocating memory for the new task
 pub fn new_task() !*os.thread.Task {
   return task_alloc.create(os.thread.Task);
 }
 
+/// Creating a new task that calls a function call
+pub fn make_task(func: anytype, args: anytype) !void {
```
Comment (N00byEdge) [inline src/thread/scheduler.zig:25]: The comment should probably say you need to call this from an existing task

File: src/thread/scheduler.zig [hunk 2]
```diff
@@ -1,31 +1,67 @@
...
+
+const task_alloc = os.memory.vmm.backed(.Ephemeral);
+var balancer_lock = os.thread.Spinlock{};
+
+/// Allocating memory for the new task
 pub fn new_task() !*os.thread.Task {
   return task_alloc.create(os.thread.Task);
 }
 
+/// Creating a new task that calls a function call
```
Comment (N00byEdge) [inline src/thread/scheduler.zig:24]: The comment should probably say you need to call this from an existing task

File: src/thread/mutex.zig [hunk 3]
```diff
@@ -25,9 +25,7 @@ pub const Mutex = struct {
   }
 
   pub fn lock(self: *@This()) void {
-    while(!self.queue.sleep(lock_impl, .{self})) {
-
-    }
+    _ = self.queue.sleep(lock_impl, .{self});
```
Comment (N00byEdge) [inline src/thread/mutex.zig:28]: This is broken. This tries to lock once. If failed, it just returns.

**Review by N00byEdge — CHANGES_REQUESTED** [review]
Why delete semaphore and condvar? I don't see why they would remain unused. For example, a semaphore for waiting for smp to start. They're useful primitives, hook into the multitasking system, and they're familiar for people just joining the project. These are intended for usage within a system, while message passing would make sense for communications between them.

File: src/thread/scheduler.zig (hunk 2 above)
Comment (notYuriy) [reply src/thread/scheduler.zig:24]: Why? It is called from pretty much alive tasks, see AHCI

Comment (N00byEdge) [reply src/thread/scheduler.zig:24]: Existing, not exiting

**Comment by notYuriy** [comment]
> Why delete semaphore and condvar? I don't see why they would remain unused. For example, a semaphore for waiting for smp to start. They're useful primitives, hook into the multitasking system, and they're familiar for people just joining the project. These are intended for usage within a system, while message passing would make sense for communications between them.

SMP is started in the bootloader (stivale1 rip PR incoming). They won't be used in the core kernel and if they are going to be used in the other system components, that would require a different implementation to make those primitives async ready (instead of just blocking the current thread)

**Comment by N00byEdge** [comment]
> They won't be used in the core kernel
Why not? That's what I just said they may. What I meant is the SMP handoff from stivale2.

> if they are going to be used in the other system components, that would require a different implementation to make those primitives async ready (instead of just blocking the current thread)
I would argue they already are. If you block on one, you're yanked out of the scheduler queue and don't get scheduled until you're done. That's what the WaitQueue is for.

**Comment by notYuriy** [comment]
> I would argue they already are. If you block on one, you're yanked out of the scheduler queue and don't get scheduled until you're done. That's what the WaitQueue is for.
That is being async-ready in the system-wide context. However, once tasks will actually become isolated units and belong to different processes (maybe even run preemptively), those primitives should be async in the context of one task: a thread should be able to lock on a mutex, semaphore, or a condvar, such that instead of blocking completely, it can continue doing its work and wait for the notification in the notification queue.

**Comment by notYuriy** [comment]
We should not forget that we have full git history and in case some of this code is needed, we can return and grab it from there. Otherwise, it's just a dead code that serves no purpose.

**Review by N00byEdge — APPROVED** [review]

**Pull request merged**